sdks/python/apache_beam/typehints/arrow_type_compatibility.py

```diff
@@ -0,0 +1,354 @@
+  # instances, pa.map_ will only accept DataType instances. This makes it
+  # impossible to propagate nullability.
+  #
+  # Note this was changed in:
+  # https://github.com/apache/arrow/commit/64bef2ad8d9cd2fea122cfa079f8ca3fea8cdf5d
+  #
+  # Here we define a custom arrow map conversion function to handle these cases
+  # and error as appropriate.
+
+  OPTIONS_WARNING = (
+      "Beam options will not be propagated to arrow schema. If options must be "
+      "propagated please use pyarrow>=6.0.0. Dropped options\n:{option}")
+
+  def _make_arrow_map(beam_map_type: schema_pb2.MapType):
+    if beam_map_type.key_type.nullable:
+      raise TypeError('Arrow map key field cannot be nullable')
+    elif beam_map_type.value_type.nullable:
+      raise TypeError(
+          "pyarrow<6 does not support creating maps with nullable "
+          "values. Please use pyarrow>=6.0.0")
+
+    return pa.map_(
+        _arrow_type_from_beam_fieldtype(beam_map_type.key_type),
+        _arrow_type_from_beam_fieldtype(beam_map_type.value_type))
+
+  def _arrow_map_to_beam_map(arrow_map_type):
+    return schema_pb2.MapType(
+        key_type=_beam_fieldtype_from_arrow_type(arrow_map_type.key_type),
+        value_type=_beam_fieldtype_from_arrow_type(arrow_map_type.item_type))
+
+else:
+
+  def _make_arrow_map(beam_map_type: schema_pb2.MapType):
+    return pa.map_(
+        _arrow_field_from_beam_fieldtype(beam_map_type.key_type),
+        _arrow_field_from_beam_fieldtype(beam_map_type.value_type))
+
+  def _arrow_map_to_beam_map(arrow_map_type):
+    return schema_pb2.MapType(
+        key_type=_beam_fieldtype_from_arrow_field(arrow_map_type.key_field),
+        value_type=_beam_fieldtype_from_arrow_field(arrow_map_type.item_field))
+
+
+def _arrow_type_from_beam_fieldtype(
+    beam_fieldtype: schema_pb2.FieldType,
+) -> Tuple[pa.DataType, Optional[Dict[bytes, bytes]]]:
+  # Note this function is not concerned with beam_fieldtype.nullable, as
+  # nullability is a property of the Field in Arrow.
+  type_info = beam_fieldtype.WhichOneof("type_info")
+  if type_info == 'atomic_type':
+    try:
+      output_arrow_type = ATOMIC_TYPE_TO_PYARROW[beam_fieldtype.atomic_type]
+    except KeyError:
+      raise ValueError(
+          "Unsupported atomic type: {0}".format(beam_fieldtype.atomic_type))
+  elif type_info == "array_type":
+    output_arrow_type = pa.list_(
+        _arrow_field_from_beam_fieldtype(
+            beam_fieldtype.array_type.element_type))
+  elif type_info == "map_type":
+    output_arrow_type = _make_arrow_map(beam_fieldtype.map_type)
+  elif type_info == "row_type":
+    schema = beam_fieldtype.row_type.schema
+    # Note schema id and options are handled at the arrow field level, they are
+    # added at field-level metadata.
+    output_arrow_type = pa.struct(
+        [_arrow_field_from_beam_field(field) for field in schema.fields])
+  elif type_info == "logical_type":
+    raise NotImplementedError()
```

Comment:
Why is this not implemented?

Reply:
It's not necessary for an MVP. I populated the error message and added an issue/TODO